ScriptAnalyzer2/Execution/RuleExecutor.cs

```diff
@@ -0,0 +1,150 @@
+﻿using Microsoft.PowerShell.ScriptAnalyzer.Rules;
+using System.Collections.Generic;
+using System.Linq;
+using System.Management.Automation.Language;
+
+namespace Microsoft.PowerShell.ScriptAnalyzer
+{
+    public interface IRuleExecutor
+    {
+        void AddRule(ScriptRule rule);
+
+        IReadOnlyCollection<ScriptDiagnostic> CollectDiagnostics();
+    }
+
+    internal class SequentialRuleExecutor : IRuleExecutor
+    {
+        private readonly Ast _scriptAst;
+
+        private readonly IReadOnlyList<Token> _scriptTokens;
+
+        private readonly string _scriptPath;
+
+        private readonly List<ScriptDiagnostic> _diagnostics;
+
+        public SequentialRuleExecutor(Ast ast, IReadOnlyList<Token> tokens, string scriptPath)
+        {
+            _scriptAst = ast;
+            _scriptTokens = tokens;
+            _scriptPath = scriptPath;
+            _diagnostics = new List<ScriptDiagnostic>();
+        }
+
+        public void AddRule(ScriptRule rule)
+        {
+            _diagnostics.AddRange(rule.AnalyzeScript(_scriptAst, _scriptTokens, _scriptPath));
+        }
+
+        public IReadOnlyCollection<ScriptDiagnostic> CollectDiagnostics()
+        {
+            return _diagnostics;
+        }
+    }
+
+    internal class ParallelLinqRuleExecutor : IRuleExecutor
+    {
+        private readonly Ast _scriptAst;
+
+        private readonly IReadOnlyList<Token> _scriptTokens;
+
+        private readonly string _scriptPath;
+
+        private readonly List<ScriptRule> _parallelRules;
+
+        private readonly List<ScriptRule> _sequentialRules;
+
+        public ParallelLinqRuleExecutor(Ast scriptAst, IReadOnlyList<Token> scriptTokens, string scriptPath)
+        {
+            _scriptAst = scriptAst;
+            _scriptTokens = scriptTokens;
+            _scriptPath = scriptPath;
+            _parallelRules = new List<ScriptRule>();
+            _sequentialRules = new List<ScriptRule>();
+        }
+
+        public void AddRule(ScriptRule rule)
+        {
+            if (rule.RuleInfo.IsThreadsafe)
+            {
+                _parallelRules.Add(rule);
+                return;
+            }
+
+            _sequentialRules.Add(rule);
+        }
+
+        public IReadOnlyCollection<ScriptDiagnostic> CollectDiagnostics()
+        {
+            List<ScriptDiagnostic> diagnostics = _parallelRules.AsParallel()
+                .SelectMany(rule => rule.AnalyzeScript(_scriptAst, _scriptTokens, _scriptPath))
+                .ToList();
+
+            foreach (ScriptRule sequentialRule in _sequentialRules)
+            {
+                diagnostics.AddRange(sequentialRule.AnalyzeScript(_scriptAst, _scriptTokens, _scriptPath));
+            }
+
+            return diagnostics;
+        }
+    }
+
+    /*
+    internal class DataflowRuleExecutor : IRuleExecutor
```

Comment:
Yeah, this didn't work initially but I'm hoping to keep it to work on to see if we can improve it. Alternatively, it might be easier to convert this to a Channel now